File: 0008-string-to-integer-atoi/solution.c
int myAtoi(char* a) {
    int i = 0;
    int sign = 1;
    long res = 0;
    while (a[i] == ' ')
        i++;
    if (a[i] == '+' || a[i] == '-') {
        if (a[i] == '-') {
            sign = -1;
        }
        i++;
    }
    while (a[i] >= '0' && a[i] <= '9') {
        res = res * 10 + (a[i] - '0');
        if (sign == 1 && res > INT_MAX) {
            return INT_MAX;
        } else if (sign == -1 && -res < INT_MIN) {
            return INT_MIN;
        }
        i++;
    }
    return (int)(sign * res);
}

File: 0008-string-to-integer-atoi/solution.java
class Solution {
    public int myAtoi(String s) {
       int idx = 0, n = s.length(), sign = 1, res = 0;

       while(idx < n && s.charAt(idx) == ' '){
            idx++;
       }

       if(idx < n && (s.charAt(idx) == '-' || s.charAt(idx) == '+')){
            sign = (s.charAt(idx) == '-') ? -1 : 1;
            idx++;
       }

       while(idx < n && Character.isDigit(s.charAt(idx))){
            int digit = s.charAt(idx) - '0';

            if(res > (Integer.MAX_VALUE - digit) / 10){
                return (sign == 1) ? Integer.MAX_VALUE :Integer.MIN_VALUE;
            }

            res = res * 10 + digit;
            idx ++;
       }
       return res * sign;
    }
}
